flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogPropertiesUtil.java

```diff
@@ -0,0 +1,470 @@
+        } catch (Exception e) {
+            throw new CatalogException("Error in deserializing catalog table.", e);
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Helper methods and constants
+    // --------------------------------------------------------------------------------------------
+
+    private static final String SCHEMA = "schema";
+
+    private static final String NAME = "name";
+
+    private static final String DATA_TYPE = "data-type";
+
+    private static final String EXPR = "expr";
+
+    private static final String METADATA = "metadata";
+
+    private static final String VIRTUAL = "virtual";
+
+    private static final String PARTITION_KEYS = "partition.keys";
+
+    private static final String WATERMARK = "watermark";
+
+    private static final String WATERMARK_ROWTIME = "rowtime";
+
+    private static final String WATERMARK_STRATEGY = "strategy";
+
+    private static final String WATERMARK_STRATEGY_EXPR = WATERMARK_STRATEGY + '.' + EXPR;
+
+    private static final String WATERMARK_STRATEGY_DATA_TYPE = WATERMARK_STRATEGY + '.' + DATA_TYPE;
+
+    private static final String PRIMARY_KEY_NAME = "primary-key.name";
+
+    private static final String PRIMARY_KEY_COLUMNS = "primary-key.columns";
+
+    private static final String COMMENT = "comment";
+
+    private static Map<String, String> deserializeOptions(Map<String, String> map) {
+        return map.entrySet().stream()
+                .filter(
+                        e -> {
+                            final String key = e.getKey();
+                            return !key.startsWith(SCHEMA + '.')
+                                    && !key.startsWith(PARTITION_KEYS + '.')
+                                    && !key.equals(COMMENT);
+                        })
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static List<String> deserializePartitionKeys(Map<String, String> map) {
+        final int partitionCount = getCount(map, PARTITION_KEYS, NAME);
+        final List<String> partitionKeys = new ArrayList<>();
+        for (int i = 0; i < partitionCount; i++) {
+            final String partitionNameKey = PARTITION_KEYS + '.' + i + '.' + NAME;
+            final String partitionName = getValue(map, partitionNameKey);
+            partitionKeys.add(partitionName);
+        }
+        return partitionKeys;
+    }
+
+    private static Schema deserializeSchema(Map<String, String> map) {
+        final Builder builder = Schema.newBuilder();
+
+        deserializeColumns(map, builder);
+
+        deserializeWatermark(map, builder);
+
+        deserializePrimaryKey(map, builder);
+
+        return builder.build();
+    }
+
+    private static void deserializePrimaryKey(Map<String, String> map, Builder builder) {
+        final String constraintNameKey = SCHEMA + '.' + PRIMARY_KEY_NAME;
```

Comment:
I think it's worth extracting the separator into a constant since it's widely used here. Maybe even just a small utility function that takes a list of strings and concatenates them with the separator in between (i.e., essentially calling `String.join(SEPARATOR, …)`